### PR TITLE
Infra/masked input to ts

### PR DESCRIPTION
**Pull request by ethanshar**
## Description
Remove usage of legacy TextField from MaskedInput and migrate component to TS and function component

## Changelog
Refactor MaskedInput component, remove legacy TextField usage

### CODE REVIEW

**Review by M-i-k-e-l — COMMENTED** [review]
Consider moving the screen to TS as well.
Also, the screen did not work as expected for me, typing more than one char in the first example did not always work.

File: src/components/maskedInput/index.tsx [hunk 1]
```diff
@@ -1,90 +1,111 @@
...
+  useEffect(() => {
+    if (props.value !== value) {
+      setValue(props.value);
+    }
+  }, [props.value]);
+
+  useEffect(() => {
+    keyboardDidHideListener.current = Keyboard.addListener('keyboardDidHide', () => {
+      if (inputRef.current?.isFocused?.()) {
+        inputRef.current?.blur?.();
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/components/maskedInput/index.tsx:62]: Why not? It's safer, no?

File: src/components/maskedInput/index.tsx [hunk 2]
```diff
@@ -1,90 +1,111 @@
...
+function MaskedInput(props: MaskedInputProps, ref: ForwardedRef<any>) {
+  const {containerStyle, renderMaskedText, onChangeText, ...others} = props;
+  const [value, setValue] = useState(props.value);
+  const inputRef = useRef<TextInput>();
+  const keyboardDidHideListener = useRef<any>();
+
+  useImperativeHandle(ref, () => {
+    return {
+      isFocused: () => inputRef.current?.isFocused(),
+      focus: () => inputRef.current?.focus(),
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/components/maskedInput/index.tsx:42]: I'm not sure what you mean..

File: src/components/maskedInput/index.tsx [hunk 3]
```diff
@@ -1,90 +1,111 @@
...
-import React from 'react';
-import {StyleSheet, ViewPropTypes, Keyboard} from 'react-native';
-import BaseInput from '../baseInput';
-import TextField from '../textField';
+import {StyleSheet, Keyboard, TextInput, TextInputProps, StyleProp, ViewStyle} from 'react-native';
 import View from '../view';
 import Text from '../text';
 import TouchableOpacity from '../touchableOpacity';
 
+export interface MaskedInputProps extends TextInputProps {
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/components/maskedInput/index.tsx]: I think that extending TextField, the old or the new is creating unnecessary complexity for this component. 
But I see your point, so I created a middleware migrator file and added a migrate prop, so users can move to the new implementation when they are ready. 

For usages in One App, no one uses any of the TextField features

**Review by M-i-k-e-l — COMMENTED** [review]
I could not comment on older conversations so I had to re-create them :/

**Comment by M-i-k-e-l** [comment]
> Consider moving the screen to TS as well. 

I think you might have missed this.

> Also, the screen did not work as expected for me, typing more than one char in the first example did not always work.

I think that using the computer's keyboard is the issue, using the simulator \ Android works.

File: src/components/maskedInput/new.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  useEffect(() => {
+    if (props.value !== value) {
+      setValue(props.value);
+    }
+  }, [props.value]);
+
+  useEffect(() => {
+    keyboardDidHideListener.current = Keyboard.addListener('keyboardDidHide', () => {
+      if (inputRef.current?.isFocused?.()) {
+        inputRef.current?.blur?.();
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/components/maskedInput/new.tsx]: Fixed

File: src/index.ts [hunk 5]
```diff
@@ -86,7 +86,6 @@ export {default as Icon, IconProps} from './components/icon';
 export {default as Image, ImageProps} from './components/image';
 export {default as ListItem, ListItemProps} from './components/listItem';
 export {default as LoaderScreen, LoaderScreenProps} from './components/loaderScreen';
-// @ts-expect-error
 export {default as MaskedInput} from './components/maskedInput';
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/index.ts]: Exported

**Comment by ethanshar** [comment]
@M-i-k-e-l 
I fixed the PR comments
I also changed a little the "new" implementation after playing with in the private components. 
First, I turned it into an uncontrolled component because of how the `formatter` (used to be called `transformer`) used to work. So I also added an `initialValue` prop

File: src/components/maskedInput/new.tsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+  const _onChangeText = useCallback((value: string) => {
+    const formattedValue = formatter(value) ?? '';
+    setValue(formattedValue);
+    onChangeText?.(formattedValue);
+  },
+  [onChangeText, formatter]);
+
+  const focus = useCallback(() => {
+    inputRef.current?.focus?.();
```
Comment (M-i-k-e-l) [inline src/components/maskedInput/new.tsx:85]:
```suggestion
    inputRef.current?.focus();
```

File: src/components/maskedInput/new.tsx [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  ReactElement,
+  ForwardedRef
+} from 'react';
+import _ from 'lodash';
+import {StyleSheet, Keyboard, TextInput, TextInputProps, StyleProp, ViewStyle} from 'react-native';
+import View from '../view';
+import Text from '../text';
+import TouchableOpacity from '../touchableOpacity';
+
+export interface MaskedInputProps extends TextInputProps {
```
Comment (M-i-k-e-l) [inline src/components/maskedInput/new.tsx:17]:
```suggestion
export interface MaskedInputProps extends Omit<TextInputProps, 'value'> {
```

File: src/components/maskedInput/new.tsx [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+   */
+  containerStyle: StyleProp<ViewStyle>;
+}
+
+/**
+ * @description: Mask Input to create custom looking inputs with custom formats
+ * @gif: https://camo.githubusercontent.com/61eedb65e968845d5eac713dcd21a69691571fb1/68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f4b5a5a7446666f486f454b334b2f67697068792e676966
+ * @example: https://github.com/wix/react-native-ui-lib/blob/master/demo/src/screens/componentScreens/MaskedInputScreen.js
+ */
+function MaskedInput(props: MaskedInputProps, ref: ForwardedRef<any>) {
```
Comment (M-i-k-e-l) [inline src/components/maskedInput/new.tsx:41]: I've written this in the main PR area, they should make that more noticeable somehow...
> Consider moving the screen to TS as well.

File: src/components/maskedInput/new.tsx [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  const {initialValue, formatter = _.identity, containerStyle, renderMaskedText, onChangeText, ...others} = props;
+  const [value, setValue] = useState(initialValue);
```
Comment (M-i-k-e-l) [inline src/components/maskedInput/new.tsx:43]: Do we want to format the `initialValue` here?

File: src/components/maskedInput/new.tsx [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        inputRef.current?.clear();
+        setValue('');
+        // NOTE: This fixes an RN issue - when triggering imperative clear method, it doesn't call onChangeText
+        onChangeText?.('');
+      }
+    };
+  });
+
+  useEffect(() => {
+    if (initialValue !== value) {
```
Comment (M-i-k-e-l) [inline src/components/maskedInput/new.tsx:62]: Do we want to format the `initialValue` here?

Comment (ethanshar) [reply src/components/maskedInput/new.tsx:62]: I guess the same answer as the other comment

File: src/components/maskedInput/new.tsx (hunk 8 above)
Comment (ethanshar) [reply src/components/maskedInput/new.tsx:41]: Definitely. Fixed and also add a new cool example (:

File: demo/src/screens/componentScreens/MaskedInputScreen.tsx [hunk 11]
```diff
@@ -56,6 +54,20 @@ export default class MaskedInputScreen extends Component {
     );
   }
 
+  renderPINCode = (value = '') => {
+    return (
+      <View row centerH>
+        {_.times(4, i => {
+          return (
+            <View marginR-s3 center style={styles.pinCodeSquare}>
```
Comment (M-i-k-e-l) [inline demo/src/screens/componentScreens/MaskedInputScreen.tsx:62]: Missing `key`

File: demo/src/screens/componentScreens/MaskedInputScreen.tsx [hunk 12]
```diff
@@ -85,11 +97,24 @@ export default class MaskedInputScreen extends Component {
...
             keyboardType={'numeric'}
           />
+
+          <Text text70 marginT-s5 marginB-s4>
+            PIN Code
+          </Text>
+          <MaskedInput
+            migrate
+            maxLength={4}
+            ref={this.priceInput}
```
Comment (M-i-k-e-l) [inline demo/src/screens/componentScreens/MaskedInputScreen.tsx:112]: You should create a dedicated `ref`, currently only one of the two users (price \ pin) are deleted.

File: demo/src/screens/componentScreens/MaskedInputScreen.tsx (hunk 11 above)
Comment (ethanshar) [reply demo/src/screens/componentScreens/MaskedInputScreen.tsx]: right, fixed!

File: demo/src/screens/componentScreens/MaskedInputScreen.tsx (hunk 12 above)
Comment (ethanshar) [reply demo/src/screens/componentScreens/MaskedInputScreen.tsx]: Fixed as well (:

**Review by M-i-k-e-l — APPROVED** [review]

**Pull request merged**